### PR TITLE
LPS-190596 | Allow users to define sorts

### PR DESCRIPTION
**Pull request by Hijky**
Background:
Add a testcase to export object entries at instance level

Test Plan
SortingIsPresentInEndpointConfiguration
When navigating to the endpoint Configuration tab
Then Sorting box to add a sort using OData is present

SortingWorksWithSchemaPropertiesMadeOfMainObjectFields
When setting the Sorting to 'age:asc'
Then students appear in ascending order in the endpoint response

CanResetSorting
And Given setting the Sorting to 'age:desc'
And Given students appear in descending order of student_age values in the endpoint response
When I empty the Sorting box in endpoint configuration
Then Student entries appear in the default order

SortingAndSortParameterReturnTheSameData
And Given setting the Sorting to 'age:asc'
And Given endpoint returning student in ascending order of age
When I empty the Sorting
And When I call endpoint with '?sort=student_age:asc' in API Explorer
Then endpoint returns student in ascending order of age

CanApplySortingAndSortParameter
And Given setting the Sorting to 'age:asc'
When I call endpoint with '?sort=age:desc' in API Explorer
Then endpoint returns student in descemding order of student_age values

SortingWorksWithMainObjectFieldsThatAreNotPartOfSchemaProperties
When setting the Sorting to 'dateCreated:asc'
Then endpoint returns students in order of creation listed as student_age: '0', '25', '101', '-1'

Jira ticket:
https://liferay.atlassian.net/browse/LPS-198371

### CODE REVIEW

**Comment by magjed4289** [comment]
ci:close

**Comment by magjed4289** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SchemaAPI.macro [hunk 1]
```diff
@@ -42,9 +42,14 @@ definition {
 	macro createAPISchema {
 		Variables.assertDefined(parameterList = "${mainObjectDefinitionERC},${name},${externalReferenceCode}");
 
+		if (!(isSet(externalReferenceCodeSchema))) {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SchemaAPI.macro:45]: Could you rename to schemaExternalReferenceCode?

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase [hunk 2]
```diff
@@ -54,7 +54,7 @@ definition {
 		property portal.acceptance = "true";
 
 		task ("And Given an published API application with endpoint with related schema with Student as the main object") {
-			var response = ApplicationAPI.createAPIApplication(
+			ApplicationAPI.createAPIApplication(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:57]: :+1:

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+				en_US_label = "student",
+				en_US_plural_label = "students",
+				name = "Student",
+				objectDefinitionExternalReferenceCode = "student",
+				requiredStringFieldName = "age",
+				requiredStringFieldNameExternalReferenceCode = "studentAge");
+		}
+
+		task ("And Given Student entries with age '0', '25', '101', '-1' created") {
+			CustomObjectAPI.createObjectEntryWithFields(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase:27]: Could you try a for loop to create these 4 entries at the same time?

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+				fieldName = "age",
+				fieldValue = 101);
+
+			CustomObjectAPI.createObjectEntryWithFields(
+				en_US_plural_label = "students",
+				fieldName = "age",
+				fieldValue = "-1");
+		}
+
+		task ("Given ApiApplication with an endpoint of scope Company created and published") {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase:48]: And Given?

Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase:48]: mum...I copy from the ticket, I will update this

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase [hunk 5]
```diff
@@ -0,0 +1,299 @@
+@component-name = "portal-headless"
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase:1]: @magjed4289 The folder name(name of epic ticket) seems a little long. For these kinds of epic tickets, are we going to use a shorter name, for example `[UI]AllowUsersToDefinePrefiltersAndSorts`? cc @Hijky

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SchemaAPI.macro (hunk 1 above)
Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SchemaAPI.macro:45]: I checked we also named the ERC like this before, so I think it's better to use this

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro [hunk 6]
```diff
@@ -215,6 +234,22 @@ definition {
 		Click(locator1 = "APIBuilder#EDIT_BUTTON");
 	}
 
+	macro goToEditRelatedEntryInEditApplication {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro:237]: The value you defined is {endpointName}, better to update to a proper macro name

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro [hunk 7]
```diff
@@ -162,6 +162,25 @@ definition {
 		}
 	}
 
+	macro editAndPublishRelatedEntryConfigInEditApplication {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro:165]: How about naming it to be addAndPublishFiltersAndSortingInEndpointConfiguration?

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
ci:stop

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase (hunk 5 above)
Comment (magjed4289) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]AllowUsersToDefinePrefiltersAndSortsToRetrieveOnlyASelectionOfEntriesInThePreferredOrder/AllowUsersToDefineSorts.testcase:1]: Yeah, we can definitely shorten it to [UI]AllowUsersToDefinePrefiltersAndSorts, for the rest of the cases in the future we can think of some changes like operating wth LPS or ayn other idea. Let's see what the tendency with long names will be ;D

**Review by magjed4289 — APPROVED** [review]

**Comment by Hijky** [comment]
ci:stop

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
ci:forward

**Comment by Hijky** [comment]
ci:reopen

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Comment by Hijky** [comment]
The failures are about 'environment is not locked', it seems CI error.

**Comment by Hijky** [comment]
ci:forward:force

**Pull request closed without merging**